articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+
+This problem can be solved in two ways:
+- Having conventions to match DDS QoS profiles names to ROS entities.
+  This has been proposed in https://github.com/ros2/rmw_fastrtps/pull/335, and could work well for most use cases.
+- Adding a rmw implementation specific "payload" that it's then used by the rmw implementation.
+  In DDS based implementations, this payload can be just the profile name.
+
+As an example of the second alternative:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">
+            <qos>
+                <rmw_payload>
+                    <profile_name>my_qos_profile_name</profile_name>
+                </rmw_payload>
+            </qos>
+        </publisher>
+    </node>
+</qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            rmw_payload:
+                profile_name: my_qos_profile_name
+```
+
+For this to work, `rmw_qos_profile_t` will need to be extended with a `rmw_qos_profile_payload_t` member defined like:
+
+```c
+typedef struct rmw_qos_profile_payload_item_t {
+  char * key;
+  char * value;
+} rmw_qos_profile_payload_item_t;
+
+typedef struct rmw_qos_profile_payload_t {
+  rmw_qos_profile_payload_item_t * key_value_pairs;
+} rmw_qos_profile_payload_t;
+```
+
+It would be up to the implementation what to do with the passed key value pairs.
+
+Note:
+  The proposed API is limited, if we would want for example to define vendor specific QoS inline in the ROS QoS profile file.
+  If YAML format is used, we could take advantage of it's natural representation of integers,floats,etc, instead of forcing everything to be an string.
+  The complexity of this approach doesn't seem to be worth it.
+Note 2:
+  The rmw specific payload could simply be an string, instead of key value pairs.
+  In that case, the `profile_name` sub-tag would be deleted and `my_qos_profile_name` would be the content of the string.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+The node options could have a flag to ignore the new `--qos-file` argument.
+In a similar way, there could be an option for publishers/subscriptions/clients and services to allow loading the QoS from an external source (side load).
+
+In nodes designed to be reused, it does make sense to allow side-loading the qos settings of all entities.
```

Comment:
This sounds a bit like a "loaded" sentence. I would argue that most nodes are designed to be reused, but this does not imply that all entities should allow their QoS to be configurable (e.g. going back to my static TF broadcaster example). I would rephrase this to something like

> There are many cases where it makes sense to allow external configuration of QoS settings for all entities.

Reply:
I'm referring to the entities that the node's author created explicitly.
For the ones that are created behind the scenes by other classes, we should add in them a callback or opt-out the feature.

Reply:
Rephrased in https://github.com/ros2/design/pull/296/commits/08294d9413d19d64d0c77cecc3f22a7937e7fc57.